python/graphstorm/distiller/distiller.py

```diff
@@ -0,0 +1,351 @@
+                        batch["labels"])
+                    total_loss += loss.item()
+                    batch_index += 1
+            index += 1
+
+        mean_total_loss = total_loss / batch_index
+        print (f"Eval MSE at step {global_step}: {mean_total_loss}")
+        model.train()
+        eval_dataset_provider.release_iterator()
+
+    def train_shard(
+        self, 
+        global_step, 
+        model, 
+        optimizer, 
+        train_dataset_provider, 
+        eval_dataset_provider, 
+        saved_path,
+        save_model_frequency,
+        eval_frequency,
+        max_global_step,
+    ):
+        """
+        Train using one shard from train_dataset_provider.
+        Parameters
+        ----------
+        global_step : int
+            Global step of the model checkpoint.
+        model : GSDistilledModel
+            Distilled student model.
+        optimizer : torch optimizer
+            optimizer for distillation.
+        train_dataset_provider : DataProvider
+            Data provider for training data.
+        eval_dataset_provider : DataProvider
+            Data provider for validation data.
+        saved_path : str
+            Path to save the model.
+        save_model_frequency : int,
+            Interval to save the model checkpoint.
+        eval_frequency : int,
+            Interval for evaluation.
+        max_global_step : int,
+            Maximum steps for distillation training.
+    
+        Returns
+        -------
+        bool : whether to stop distillation.
+        int : Global step of the model checkpoint.
+        """
+        dataset_iterator = train_dataset_provider.get_iterator()
+        if not dataset_iterator:
+            raise RuntimeError("No training data. Check training dataset as the data sampler"
+                                    "is supposed to be infinite.")
+
+        # TODO (HZ): support prefetch to speed up the training
+        model.train()
+        shard_loss = 0
+        complete = False
+
+        for batch_num, batch in enumerate(dataset_iterator):
+            try:
+                batch = self.to_device(batch, self.device)  # Move to device
+
+                loss = model(batch["input_ids"],
+                    batch["attention_mask"],
+                    batch["labels"])
+
+                shard_loss += loss.item()
+                optimizer.zero_grad()
+                loss.backward()
+                optimizer.step()
+                mean_shard_loss = shard_loss / (batch_num + 1)
+                if global_step % 20 == 0:
+                    print (
+                        f"Loss for shard {train_dataset_provider.get_iterator_name()}"
+                        f" at step {global_step} = {mean_shard_loss}"
+                        f" from trainer {self.rank}"
+                    )
+
+                if global_step % save_model_frequency == 0 and global_step != 0:
+                    self.save_student_model(model, saved_path, global_step)
```

Comment:
You need to have some logic as https://github.com/awslabs/graphstorm/blob/0fe8da0beccc5a82d021155158c95c21d9840032/python/graphstorm/trainer/gsgnn_trainer.py#L222. Otherwise, you will have to many checkpoints and explode your disk.